Refactoring

BalanceBalls opened the pull request:
Alright, that's a big one :) but fear not, most of the code has just been moved with some tweaks and fixes here and there

### Why

* Before adding new features I would like to establish some architecture to split parts of the app into clearly defined parts.
This will reduce side effects and bugs upon adding new features and generally will make adding and supporting features easier.
* Some core things like window scaling and design things lacked structure. Implemented a dimension calculations module for UI elements and themes support

### Concept
The concept of the refactoring is establishing a hierarchy of elements:
- views host and manage panes , and panes host and manage components
 - elements are supposed to be isolated and to communicate using events (to avoid 'leaky' abstractions) 
<img width="803" alt="image" src="https://github.com/user-attachments/assets/8a5d855d-1982-4ef0-bf27-51ad9e113ac7" />


### What's changed
 * all core parts were decoupled from main.go into separate panes that communicate via events (tea.Msg)
 * main.go is now just an entrypoint to kickstart the main-view model
 * main-view model houses the panes and manages inter-pane communications 
 * all the size calculations were moved into a single 'dimensions' 
 * style colors were upgraded to themes that can be switched from config.json (also added a new 'blue' theme)
 * playground code has been removed

### New features
 * Added spinner during models loading
 * Made models loading async
 * Implemented async loading for panes that need IO calls to initialize
 * Themes support
 * Added cache for models
 * Added filtering to display chat models only. Filter removes image gen models, voice models etc. (Filtering added for OpenAI and Mistral)
 * Seamless switching between providers:
  -when changing inference provider (url in config) , the models will be displayed for the current provider
 * Auto zen-mode activation for extremely narrow terminal window sizes (threshold is 120)
 * Long list items trimming for narrow windows scales
 * Some tweaks and improvements to UI
 
 ### Bug fixes
  * Session list ESC key handling
  * Error events handling
  * Inconsistent scaling behavior
  
  ### New theme
 
<img width="1355" alt="image" src="https://github.com/user-attachments/assets/f6b3cd7a-ed74-4afc-b3b8-1fe04b45c63e" />

  ### P.S.
  I've done refactoring in multiple PRs. Just for reference:
  - [Chat pane refactoring](https://github.com/BalanceBalls/chatgpt-tui/pull/1)
  - [Settings initialization refactoring and code cleanup](https://github.com/BalanceBalls/chatgpt-tui/pull/2)
  - [Prompt pane refactoring](https://github.com/BalanceBalls/chatgpt-tui/pull/3)
  - [Sessions pane refactoring](https://github.com/BalanceBalls/chatgpt-tui/pull/4)
  - [Settings pane refactoring](https://github.com/BalanceBalls/chatgpt-tui/pull/5)
  - [Main view refactoring, themes, UI improvements](https://github.com/BalanceBalls/chatgpt-tui/pull/6)

BalanceBalls commented:
> Holy crow brother... thanks for doing this! I picked up Go as a hobby last year and have regrettably not been writing it a lot this year and I'm super excited to read this over.
> 
> Side note, are you on the Charm Discord server? We can chat over there as well if you ever want!
> 
> https://charm.sh/chat/

Thanks! Joined the server.

I'm not a go dev myself either, so I think we're in the same boat here :)

tearingItUp786 commented on migrations/00002_settings.sql:
```diff
@@ -8,8 +8,8 @@ CREATE TABLE settings (
-(settings_model, settings_max_tokens, settings_frequency) 
-VALUES ('gpt-3.5-turbo', 300.0, 0.0); -- Assuming you want to store 300 as a floating-point number
+(settings_id, settings_model, settings_max_tokens, settings_frequency) 
+VALUES (0, 'gpt-3.5-turbo', 300.0, 0.0); -- Assuming you want to store 300 as a floating-point number
```
🪡 Nitpick: we could probably change the default model to 4.0 eh?

tearingItUp786 commented on util/providers.go:
```diff
@@ -0,0 +1,97 @@
+
+var (
+	openAiChatModelsPrefixes = []string{"gpt-", "o1", "o3"}
+	openAiExclusionKeywords  = []string{"audio", "realtime"}
+
+	mistralExclusionKeywords = []string{"pixtral", "embed"}
```
🤔 Question: these are excluded as they aren't usable from the CLI, correct? Might be worth leaving a comment here regardless because I'm not sure myself haha.

BalanceBalls replied:
yeeeah.. you're right, kinda confusing. Left a comment in code on what it is all about. Basically `exclusion keywords` is a blacklist for models that contain any of the specified. And `prefixes` is a whitelist for models which names start with any of the specified. It affects the models list in the settings pane

tearingItUp786 commented on util/shared-colors.go:
Very nice ❤️

tearingItUp786 commented:
Okay this all makes sense to me! Before approving, would it be possible to update the `README` with the change to the config regarding the theme and letting folks know what values are allowed?

BalanceBalls commented:
> Okay this all makes sense to me! Before approving, would it be possible to update the `README` with the change to the config regarding the theme and letting folks know what values are allowed? I'll try breaking it later with some weird values after that update.
> 
> Also, it might be worth it now to add a `CONTRIBUTING.md` -- your refactor introduced a clear separation of concerns as it pertains to views and panes and having that diagram you created in this PR description in the CONTRIBUTING doc would be a great introduction for other folks (and a great reminder for me whenever I come back to this project).

Sure, updated the readme with the themes info and added a doc for contributors. Thanks!

tearingItUp786 reviewed (approved):
LGTM 🚀 . Will cut an RC with this!

BalanceBalls commented:
> LGTM 🚀 . Will cut an RC with this!

Actually, I have another PR with actual big features coming right up. Maybe we delay the release until the next PR is merged?🤔

tearingItUp786 commented:
Oops just saw this now 😓. I was struggling with the release process and just fixed it + documented it.
